modules/OFStateManager/module/src/gentable_handlers.c

```diff
@@ -277,26 +275,105 @@ ind_core_bsn_gentable_entry_add_handler(
     /* Add to table checksum */
     add_checksum(&gentable->checksum, &entry->checksum);
 
-    return;
-
 error:
     if (rv == INDIGO_ERROR_PARAM) {
         indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
                                            OF_BSN_GENTABLE_ERROR_PARAM,
-                                           err_txt);
+                                           *err_txt);
     } else if ((rv == INDIGO_ERROR_RESOURCE) ||
                (rv == INDIGO_ERROR_TABLE_FULL)) {
         indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
                                            OF_BSN_GENTABLE_ERROR_TABLE_FULL,
-                                           err_txt);
-    } else {
+                                           *err_txt);
+    } else if (rv == INDIGO_ERROR_TIME_OUT) {
+        indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
+                                           OF_BSN_GENTABLE_ERROR_UNKNOWN,
+                                           *err_txt);
+    } else if (rv < 0) {
         indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
                                            OF_BSN_GENTABLE_ERROR_UNKNOWN,
                                            "Gentable add failed");
     }
+done:
+   if (err_txt) {
+       aim_free(err_txt);
+   }
+    return;
 }
 
-void
+indigo_error_t
+ind_core_bsn_gentable_entry_add_handler(
+    of_object_t *obj,
+    indigo_cxn_id_t cxn_id)
+{
+    uint16_t table_id;
+    indigo_core_gentable_t *gentable;
+    of_list_bsn_tlv_t key, value;
+    void *priv = NULL;
+    struct ind_core_gentable_entry *entry;
+    of_desc_str_t *err_txt;
+    indigo_error_t rv = INDIGO_ERROR_NONE;
+
+    of_bsn_gentable_entry_add_table_id_get(obj, &table_id);
+
+    gentable = find_gentable_by_id(table_id);
+    if (gentable == NULL) {
+        AIM_LOG_ERROR("Nonexistent gentable id %d in gentable_entry_add message", table_id);
+        indigo_cxn_send_error_reply(
+            cxn_id, obj,
+            OF_ERROR_TYPE_BAD_REQUEST,
+            OF_REQUEST_FAILED_BAD_TABLE_ID);
+        return rv;
+    }
+
+    /* bind key and value */
+    of_bsn_gentable_entry_add_key_bind(obj, &key);
+    of_bsn_gentable_entry_add_value_bind(obj, &value);
+    err_txt = aim_zmalloc(sizeof(of_desc_str_t));
```

Comment:
i don't think we need to malloc memory for err_txt.
in case of an immediate error from the driver, err_txt should contain some useful string (as it does right now for add3/mod3/del3); in case of an error from the driver when calling the resume function, err_txt passed to the resume function called from the driver will also contain some useful string passed from the driver and should be valid until the resume function returns back to the driver.

Reply:
Just try to make add4()/mod4()/del4() can be used for original add3()/mod3()/del3() cases. Otherwise, we don't need to pass the err_txr in add4()/mod4()/del4(). What do you think?

Reply:
so add3/mod3/del3 take "of_desc_str_t err_txt" as a parameter ; i'm suggesting that add4/mod4/del4 do the same